scripts/irods/paths.py

```diff
@@ -1,139 +1,458 @@
 import grp
-import inspect
 import os
 import pwd
-
-def root_directory():
-    return os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(scripts_directory()))))
+import pathlib
+import logging
+
+try:
+    from functools import cache as func_cache
+except ImportError:
+    from functools import lru_cache as func_cache
+
+from .exceptions import IrodsError
+
+try:
+    from . import paths_cmake
+    no_paths_cmake = False
+except ImportError:
+    no_paths_cmake = True
+
+# If making changes to this file, equivalent changes may also be needed in
+# lib/core/src/irods_default_paths.cpp
+
+# TODO: consider completely transitioning to pathlib paths.
+
+_irods_install_directory_cache = None
+def _irods_install_directory():
+    global _irods_install_directory_cache
+    if _irods_install_directory_cache is None:
+        if no_paths_cmake:
+            raise IrodsError('Running from source tree. No install directories.')
+        _irods_install_directory_cache = pathlib.PurePath(os.path.normpath(paths_cmake.cmake_dir_irodshome))
+    return _irods_install_directory_cache
+
+_sysconf_install_directory_cache = None
+def _sysconf_install_directory():
```

Comment:
We should document these for others less familiar with them (i.e. me lol).

Same for `_localstate_install_directory()`. Or we reference other material that provides good explanations of these directory names.